### DISCUSSION TITLE
How to update the design of this repo to make it worth learning?

### DISCUSSION THREAD

**Discussion by vaishnavi-2901** [vaishnavi-2901/dice-game#5] [Ideas]
I am thinking of making it a live pairing game, which can be a good learning for basics of react and typescripts.
Collaborative features are good way to learn healthy basics of software engineering.

Open to suggestions!

**Comment by prashasth-nair** [comment]
You can Use WebSockets or a similar technology for real-time communication.
Allow users to join a game session and see each other's moves and scores.
You can also add leaderboard showing the scores of other players. 
But First you will need to create a database to store all the data.

**Reply by vaishnavi-2901** [reply]
you can check this repo as well:https://github.com/vaishnavi-2901/chat-application

But yes, for this would need to create a database design if we are going with multi player session. but I was thinking on the game functionality side

It is not fun to play dice, maybe we could improve there!!

**Reply by prashasth-nair** [reply]
We can add a high score system. Users can compete to beat the previous high score.

**Reply by nileshkmr99** [reply]
Hey guys, may be the game should be like this -

**Players rolling dice to get a puzzle and solve it to earn credits.**

Steps -

1. Players can join an ongoing session or start a new one.
2. Players roll the dice to determine their puzzle.
3. Based on the dice roll, they receive a puzzle to solve within a certain timeframe.
4. Successfully solving the puzzle earns the player credits.
5. The leaderboard updates in real-time to show the current standings of all players in the session.

For puzzles we can think of giving them a short educational quiz like React/Typescript/Java Quiz questions and upon correct answers they would earn credits or loss points

**Reply by prashasth-nair** [reply]
And the dice number can represent difficultly level of the puzzle. Like 1 can be easy and 6 will be the hardest.

**Reply by nileshkmr99** [reply]
@prashasth-nair sounds great but don't want to lose our initial players if they get 5s or 6s in a row 6 times 😅 , we want this game to be more engaging and fun playing with learning.

**Comment by nileshkmr99** [comment]
Can think of 2 ideas now to play this game -

1. **Players rolling dice to get a puzzle and solve it to earn credits**

Steps -

      1. Players can join an ongoing session or start a new one.
      2. Players roll the dice to determine their puzzle.
      3. Based on the dice roll, they receive a puzzle to solve within a certain timeframe.
      4. Successfully solving the puzzle earns the player credits.
      5. The leaderboard updates in real-time to show the current standings of all players in the session.

For puzzles we can think of giving them a short educational quiz like React/Typescript/Java Quiz questions and upon correct answers they would earn credits or loss points

======================================================================================

2. **Players rolling multiple die to predict a number and solve it to earn credits**

In the steps below, I am thinking of playing with 3 dice.

Steps - 

1. **Game Session**  - Players can join an ongoing session or start a new one.
2. **Roll Prediction** - Players predict the sum of the numbers that will appear on the upper faces of 3 dices after they are rolled.
3. **Roll the dice**

        - With 3 dice, possible sums range from 3 to 18.
        - Different sums provide different credit points based on their rarity.
        - Rare sums like 3 or 18 earn higher credits (e.g., 100 points).
        - Common sums like 8, 10, 12, and 13 earn fewer credits (e.g., 10-20 points, depending on the exact sum and game rules).

4. **Credits Earned** - Players earn credits based on the accuracy of their predictions.
5. **Leaderboard and Multi-Player Mode**:

      - In multi-player mode, the game is round-based.
      - Each player has 10 rounds to roll the dice and predict the outcome.
      - Correct predictions earn credits, and the leaderboard updates in real-time to show the current standings of all players in the session.

This game plan aims to balance difficulty and reward, making it exciting for players to predict less likely outcomes for higher rewards while still offering engagement for more common outcomes.

**Comment by vaishnavi-2901** [comment]
Great Ideas, guys!

but I think we can think of tweaking it’s originality. 

liked the idea of keeping numbers as levels.

and below can have multiple cards of other games.


Let’s see user story here:

1. Home page to have logo and levels and user inputs 
2. On entering user details, user can select level 
3. After level selection, user needs to select a game card to play and a link to invite others (collaborative work can be done later)
4. Some cards can be presented as multiple games like snake ladder, rope battle, car race, etc) —> can think of something else tho!
5. There we can let user play for some time and get user’s highest score only in DB.



lemme know your views.

**Reply by nileshkmr99** [reply]
That's a great idea!

Have few Questions - might come more 😅 

1. Can we have a detailed user flow diagram for the entire game from start to finish ?
2. How will the "link to invite others" work? so this will be a real-time multiplayer session game right? or player can participate at their convenience, something like async play? and also the players can join a game session at any time, or is there a specific window during which they can join?
3. How will the dice rolling mechanics integrate with the various mini-games (like snake ladder, rope battle, car race, etc.)? are we taking support of any 3rd party games for these games or we will be building in-house?
4. Also how will the scores from these various mini-games be unified or compared on the leaderboard?
5. And also will each mini-game have its own set of rules and scoring system, or will there be a some sort of unified scoring mechanism?

**Comment by nileshkmr99** [comment]
Hey @vaishnavi-2901 , need you views for building and designing this game!

So based upon on our previous discussions, I’ve been thinking about the direction for building and designing our game. So here's a detailed proposal for the user flow from start to finish, including user registration, level selection, game selection, multiplayer invitations, and scoring mechanism.

So we will be having a couple of components of our game- 

**1. Home Page**
- Here we'll display the application logo with nice design of our page and also we'll be having user input fields for registration or login.
	
	**a. User Registration and Login**
	   		- 1st time users register by providing necessary details like username or email.
	   	         - Existing users log in to access their profile and game progress.

**2. Level Selection (After login)**
- Users select a game from the available options after throwing the dice. The upper face of the dice after throwing it will decide the level of the game for the user.

**3. Game Selection** 
- Users choose a mini-game to play (e.g., snake ladder, rope battle, car race).

**4. Game Session Setup**
- Display game rules and instructions and provide options for real-time multiplayer sessions or async play.

**5. Invitation Links**
- Generate and share invitation links for multiplayer sessions and allow other players to join the game session using the link.

**6. Leaderboard**
- Track and calculate scores during gameplay, also players can take turns and interact with game elements based on dice outcomes. After that we'll compare the highest scores among players and display on the leaderboard.

**Comment by prashasth-nair** [comment]
Check Out this api to verify if an email exist or not we can use this: [Email Verification](https://eva.pingutil.com/ )

**Comment by vaishnavi-2901** [comment]
Came up with this idea:

https://docs.google.com/document/d/19YrlN07wD_aMNHDFAaIDcabIjVfzwNl_Lh6qonnhgfU/edit

Please request for access & comment for improvements.

@nileshkmr99 @prashasth-nair
